hbase-server/src/main/java/org/apache/hadoop/hbase/procedure2/store/region/RegionProcedureStore.java

```diff
@@ -0,0 +1,576 @@
+    byte[] row = Bytes.toBytes(procId);
+    mutations.add(new Put(row).addColumn(FAMILY, PROC_QUALIFIER, EMPTY_BYTE_ARRAY));
+    rowsToLock.add(row);
+  }
+
+  @Override
+  public void insert(Procedure<?> proc, Procedure<?>[] subProcs) {
+    if (subProcs == null || subProcs.length == 0) {
+      // same with update, just insert a single procedure
+      update(proc);
+      return;
+    }
+    List<Mutation> mutations = new ArrayList<>(subProcs.length + 1);
+    List<byte[]> rowsToLock = new ArrayList<>(subProcs.length + 1);
+    try {
+      serializePut(proc, mutations, rowsToLock);
+      for (Procedure<?> subProc : subProcs) {
+        serializePut(subProc, mutations, rowsToLock);
+      }
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to insert proc {}, sub procs {}", proc,
+        Arrays.toString(subProcs), e);
+      throw new UncheckedIOException(e);
+    }
+    flusherAndCompactor.onUpdate();
+  }
+
+  @Override
+  public void insert(Procedure<?>[] procs) {
+    List<Mutation> mutations = new ArrayList<>(procs.length);
+    List<byte[]> rowsToLock = new ArrayList<>(procs.length);
+    try {
+      for (Procedure<?> proc : procs) {
+        serializePut(proc, mutations, rowsToLock);
+      }
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to insert procs {}", Arrays.toString(procs), e);
+      throw new UncheckedIOException(e);
+    }
+    flusherAndCompactor.onUpdate();
+  }
+
+  @Override
+  public void update(Procedure<?> proc) {
+    try {
+      ProcedureProtos.Procedure proto = ProcedureUtil.convertToProtoProcedure(proc);
+      region.put(new Put(Bytes.toBytes(proc.getProcId())).addColumn(FAMILY, PROC_QUALIFIER,
+        proto.toByteArray()));
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to update proc {}", proc, e);
+      throw new UncheckedIOException(e);
+    }
+    flusherAndCompactor.onUpdate();
+  }
+
+  @Override
+  public void delete(long procId) {
+    try {
+      region
+        .put(new Put(Bytes.toBytes(procId)).addColumn(FAMILY, PROC_QUALIFIER, EMPTY_BYTE_ARRAY));
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to delete pid={}", procId, e);
+      throw new UncheckedIOException(e);
+    }
+    flusherAndCompactor.onUpdate();
+  }
+
+  @Override
+  public void delete(Procedure<?> parentProc, long[] subProcIds) {
+    List<Mutation> mutations = new ArrayList<>(subProcIds.length + 1);
+    List<byte[]> rowsToLock = new ArrayList<>(subProcIds.length + 1);
+    try {
+      serializePut(parentProc, mutations, rowsToLock);
+      for (long subProcId : subProcIds) {
+        serializeDelete(subProcId, mutations, rowsToLock);
+      }
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
```

Comment:
Yeah, what is the scheme here? We write all mutations as Put, then per subprocid?